pinot/inference/adlala.py

```diff
@@ -1 +1,211 @@
+# =============================================================================
+# IMPORTS
+# =============================================================================
+import torch
+
+
+# =============================================================================
+# MODULE CLASSES
+# =============================================================================
+class AdLaLa(torch.optim.Optimizer):
+    """ Adaptive Langevin-Langevin Integrator.
+
+    ref: `https://arxiv.org/abs/1908.11843`
+    
+    original problem-specific implementation:
+        `https://github.com/`
+        `TiffanyVlaar/`
+        `ThermodynamicParameterizationOfNNs/`
+        `blob/master/AdLaLa.ipynb`
+
+    Attributes
+    ----------
+    params : iterator
+        an iterator over a series of `torch.nn.Parameter` to be optimized.
+    h : float, default=0.25
+        learning rate.
+    gamma : float, dafault=0.1
+        friction parameter (for Langevin only).
+    sigma : float, default=0.01
+        driving noise amplitude (for Adaptive Langevin only).
+    epsilon: float, default=0.01
+        coupling coefficient (for Adaptive Langevin only).
+    tau: float, default=1e-4
+        temperature.
+    xi_init: float, default=1e-3
+        initial value for $xi$.
+
+    Methods
+    -------
+    step(closure): apply gradient.
+    
+
+    """
+
+    def __init__(
+            self,
+            params,
+            h=0.25,
+            gamma=0.1,
+            sigma=0.01,
+            epsilon=0.05,
+            tau=1e-4,
+            xi_init=1e-3,
+            partition='La'):
+    
+        defaults = dict(
+            h=torch.tensor(h),
+            tau=torch.tensor(tau),
+            gamma=torch.tensor(gamma),
+            sigma=torch.tensor(sigma),
+            epsilon=torch.tensor(epsilon),
+            xi_init=torch.tensor(xi_init),
+            partition='La')
+
+        super(AdLaLa, self).__init__(params, defaults)
+
+    @torch.no_grad()
+    def step(self, closure=None):
+        """ Performs a single optimization step.
+
+        Perform the steps in the following order:
+        
+        ```
+        A^1_(h/2) 
```

Comment:
You might note the initialization step(s) as well, since this is a formulation in which the weights and momenta are half a step out of sync during each step.